### PR TITLE
Improve `create_submesh` performance

### PR DESCRIPTION
**Pull request by jpdean**
The current implementation of `create_submesh` does not scale due to an `std::find` on a potentially large array in a tight loop. This PR precomputes the map to avoid this issue.

Performance results:

I ran the following code for different values of `n`:
```
msh = mesh.create_unit_cube(
        MPI.COMM_WORLD, n, n, n, ghost_mode=mesh.GhostMode.none)
tdim = msh.topology.dim
entities = mesh.locate_entities(msh, tdim, lambda x: x[0] <= 0.5)
mesh.create_submesh(msh, tdim, entities)
```
The timings are as follows:

|  | `n = 8` | `n = 16` | `n = 32` |
| ------------- | ------------- | ------------- | ------------- |
| Main  | 0.0365 s  | 0.861 s  | 46.6 s |
| This branch  | 0.00904 s  | 0.0459 s  | 0.38 s  |

The difference in performance gets larger as the mesh size increases.

### CODE REVIEW

File: cpp/dolfinx/mesh/Mesh.cpp [hunk 1]
```diff
@@ -283,16 +283,27 @@ mesh::create_submesh(const Mesh& mesh, int dim,
...
+  // NOTE: Depending on the submesh, this may be densely or sparsely
+  // populated. Is a different data structure more appropriate?
+  std::vector<int32_t> mesh_to_submesh_vertex_map(
+      mesh_vertex_index_map->size_local() + mesh_vertex_index_map->num_ghosts(),
+      -1);
+  for (std::size_t i = 0; i < submesh_to_mesh_vertex_map.size(); ++i)
+  {
+    mesh_to_submesh_vertex_map[submesh_to_mesh_vertex_map[i]] = i;
+  }
+
```
[the remark this answers is not in the document]
Comment (jpdean) [reply cpp/dolfinx/mesh/Mesh.cpp:298]: Wouldn't the offset array just be the same size as `mesh_to_submesh_vertex_map` in that case though?

[the remark this answers is not in the document]
Comment (IgorBaratta) [reply cpp/dolfinx/mesh/Mesh.cpp:298]: Yes, I think the`AdjacencyList` is only interesting when you have more than one data element per index (like a matrix).
I think a vector is fine in this case (but it also depend on how sparse your `mesh_to_submesh_vertex_map` is).

File: cpp/dolfinx/mesh/Mesh.cpp [hunk 2]
```diff
@@ -283,16 +283,27 @@ mesh::create_submesh(const Mesh& mesh, int dim,
                              * num_vertices_per_entity);
   std::vector<std::int32_t> submesh_e_to_v_offsets(1, 0);
   submesh_e_to_v_offsets.reserve(submesh_to_mesh_entity_map.size() + 1);
+
+  // Create mesh to submesh vertex map (i.e. the inverse of
+  // submesh_to_mesh_vertex_map)
+  // NOTE: Depending on the submesh, this may be densely or sparsely
+  // populated. Is a different data structure more appropriate?
+  std::vector<int32_t> mesh_to_submesh_vertex_map(
```
Comment (garth-wells) [inline cpp/dolfinx/mesh/Mesh.cpp:291]: Should be `std::int32_t`.

File: cpp/dolfinx/mesh/Mesh.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/mesh/Mesh.cpp]: A sparse alternative would be to create a sorted vector of (mesh index , submesh index) pairs, and use binary search.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**